### PR TITLE
Add TCP Transport

### PR DESCRIPTION
**Pull request by dhobbs**
Added a TCP transport, use it by adding the <transport>TCP</transport> tag to logback.xml.
Updated build to java 8.

### CODE REVIEW

**Comment by Moocar** [comment]
Thanks! I'd like to add TCP support, but I'd like to keep the JVM target at 1.6. Users shouldn't be required to upgrade their JVM to use logback-gelf. Any chance of a solution that doesn't use java 8 features?

**Comment by dpeterka** [comment]
+1 on this request

**Comment by dhobbs** [comment]
Okay, I went and actually read the graylog spec this time :)

The TCP input prefers null-terminated messages so gzipping them isn't an option (as there will almost certainly be null characters somewhere in a binary blob). So I had to add a different AppenderExecutor for TCP messages.

Re-targetted 1.6 although I don't have that version installed myself, hopefully it should be fine.

I've tested this against graylog1.0.0 and it seems to be working fine. Should be good to go.

File: src/main/java/me/moocar/logbackgelf/TcpTransport.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    private final Lock lock = new ReentrantLock();
+    private Socket socket;
+    public TcpTransport(int graylog2ServerPort, InetAddress graylog2ServerAddress) {
+        this.graylog2ServerPort = graylog2ServerPort;
+        this.graylog2ServerAddress = graylog2ServerAddress;
+    }
+
+
+    @Override
+    public void send(final byte[] data) {
```
Comment (Moocar) [inline src/main/java/me/moocar/logbackgelf/TcpTransport.java]: The functionality of the two `send` functions is identical except that the List arity takes a list. To reduce on duplication, how about this send just dispatches to `send(Arrays.aslist(data))`

File: src/main/java/me/moocar/logbackgelf/TcpTransport.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    private void handleException(Exception e) {
+        // Silently fail, same as UDP
+//        e.printStackTrace();
+    }
+
+    private void writeNullTerminated(byte[] packet, OutputStream outputStream) throws IOException {
+        byte[] nullTerminated = new byte [packet.length + 1];
+        System.arraycopy(packet, 0, nullTerminated, 0, packet.length);
+        nullTerminated[packet.length] = 0;
+        outputStream.write(nullTerminated);
```
Comment (Moocar) [inline src/main/java/me/moocar/logbackgelf/TcpTransport.java]: To save an array copy, how about just:
```java
outputStream.write(packet);
outputStream.write((byte) 0);
```

**Comment by Moocar** [comment]
The more I think about it, the more it makes sense to start Configuring the Transport in its own way. UDP and TCP have fundamentally different configuration, so deserve their own configuration sections within the logback.xml

**TCP**
- socket timeout
- keepalive?

**UDP**
- checkThreshold

Also, TCP doesn't need to chunk, so only one a single send arity is required.

I'm kind of inspired by your PR, so if you don't mind, I might take a stab at the above

File: src/main/java/me/moocar/logbackgelf/TcpTransport.java (hunk 2 above)
Comment (dhobbs) [inline src/main/java/me/moocar/logbackgelf/TcpTransport.java]: I started with that then had a paranoid thought that if there's ever multiple writers the writes could get interleaved. Doing it this way sends all the bytes to the socket together.

**Comment by dhobbs** [comment]
I agree that some of the methods don't make as much sense for the TCP transport. I left the general shape of the API intact because I'm not hugely familiar with logback and I didn't want to embark on a large change that might be harder to merge in. Feel free to use or not use any of it :)

**Comment by Moocar** [comment]
FYI, here's the approach I'm taking. Just finishing up the edges now. https://github.com/Moocar/logback-gelf/tree/the-great-appender-rejig

**Comment by dhobbs** [comment]
That's great! Much better than my hacky changes.

**Comment by Moocar** [comment]
Thanks :). I just finished the branch, any chance you'd be interested in testing it? I've done a fair bit, but it would be great to get some feedback on the new approach. You'll need to build both socket-encoder-appender and logback-gelf from scratch as they're not on maven central yet

https://github.com/Moocar/socket-encoder-appender
https://github.com/Moocar/logback-gelf/tree/the-great-appender-rejig

**Comment by Moocar** [comment]
version 2.0beta is now live (whenever maven central updates anyway). so I'm going to close this PR. Would appreciate any feedback on it. Thanks for the inspiration :)

**Pull request closed without merging**